src/goto-instrument/value_set_fi_fp_removal.cpp

```diff
@@ -0,0 +1,250 @@
+
+  // this better have the right type
+  code_typet call_type = to_code_type(function.type());
+
+  // refine the type in case the forward declaration was incomplete
+  if(call_type.has_ellipsis() && call_type.parameters().empty())
+  {
+    call_type.remove_ellipsis();
+    forall_expr(it, code.arguments())
+      call_type.parameters().push_back(code_typet::parametert(it->type()));
+  }
+
+  const exprt &pointer = to_dereference_expr(function).pointer();
+
+  // the final target is a skip
+  goto_programt final_skip;
+
+  goto_programt::targett t_final = final_skip.add_instruction();
+  t_final->make_skip();
+
+  // build the calls and gotos
+
+  goto_programt new_code_calls;
+  goto_programt new_code_gotos;
+
+  for(const auto &fun : functions)
+  {
+    // call function
+    goto_programt::targett t1 = new_code_calls.add_instruction();
+    t1->make_function_call(code);
+    to_code_function_call(t1->code).function() = fun;
+
+    // the signature of the function might not match precisely
+    const namespacet ns(goto_model.symbol_table);
+    fix_argument_types(to_code_function_call(t1->code), ns);
+    fix_return_type(
+      to_code_function_call(t1->code), new_code_calls, goto_model);
+
+    // goto final
+    goto_programt::targett t3 = new_code_calls.add_instruction();
+    t3->make_goto(t_final, true_exprt());
+
+    // goto to call
+    address_of_exprt address_of(fun, pointer_type(fun.type()));
+
+    goto_programt::targett t4 = new_code_gotos.add_instruction();
+    t4->make_goto(
+      t1,
+      equal_exprt(
+        pointer, typecast_exprt::conditional_cast(address_of, pointer.type())));
+  }
+
+  // fall-through
+  // if(add_safety_assertion)
+  {
+    goto_programt::targett t = new_code_gotos.add_instruction();
+    t->make_assertion(false_exprt());
+    t->source_location.set_property_class("pointer dereference");
+    t->source_location.set_comment("invalid function pointer");
+  }
+
+  goto_programt new_code;
+
+  // patch them all together
+  new_code.destructive_append(new_code_gotos);
+  new_code.destructive_append(new_code_calls);
+  new_code.destructive_append(final_skip);
+
+  // set locations
+  Forall_goto_program_instructions(it, new_code)
+  {
+    irep_idt property_class = it->source_location.get_property_class();
+    irep_idt comment = it->source_location.get_comment();
+    it->source_location = target->source_location;
+    if(!property_class.empty())
+      it->source_location.set_property_class(property_class);
+    if(!comment.empty())
+      it->source_location.set_comment(comment);
+  }
+
+  goto_programt::targett next_target = target;
+  next_target++;
+
+  goto_program.destructive_insert(next_target, new_code);
+
+  // We preserve the original dereferencing to possibly catch
+  // further pointer-related errors.
+  code_expressiont code_expression(function);
+  code_expression.add_source_location() = function.source_location();
+  target->code.swap(code_expression);
+  target->type = OTHER;
```

Comment:
what does an OTHER instruction even do? Maybe use an assign targeting a dummy symbol?

Reply:
An OTHER instruction is used for anything that isn't one of the other instruction types

An assign would be treated differently when symexed: OTHER will essentially be ignored, since it's not one of the specific cases handled by symex_other and ASSIGN will be handled by symex_assign, and we don't want this to be symexed.

Could just remove it since there currently isn't any code that consumes it at the moment, if that's preferred?

Reply:
What's wrong with just executing it with an assign? The comment sounds like we do want to deref the pointer "for real" to obtain any side-effects of that action (i.e., a crash if it's null) as might be obtained from later passes adding pointer-sanity assertions, and that puts it on less shaky ground than using an OTHER instruction that is liable to be ignored in a way that an ASSIGN rhs probably won't be

Reply:
How do later passes that add pointer-sanity assertions handle assigns? Or are these hypothetical future passes? Obviously we should make this work for existing passes, but I don't think it's obvious that a dummy variable is less easily ignored by a future pass writer than an "OTHER" instruction

Reply:
OK never mind, the particular pass I was thinking about was `goto_check` and that appears to expect these instructions: https://github.com/diffblue/cbmc/blob/30244047f309d73bdeef4aec9d8bcb7342f82b73/src/analyses/goto_check.cpp#L1979